Sources/Fuzzilli/Corpus/MarkovCorpus.swift

```diff
@@ -0,0 +1,206 @@
+        // Only do computationally expensive work choosing the next program when there is a solid
+        // baseline of execution data. The data tracked in the statistics module is not used, as modules are intended 
+        // to not be required for the fuzzer to function.
+        if totalExecs > 250 {
+            // Check if more programs are needed
+            if programExecutionQueue.count == 0 {
+                regenProgramList()
+            }
+            assert(programExecutionQueue.count > 0)
+            if remainingEnergy > 0 {
+                remainingEnergy -= 1
+            } else {
+                remainingEnergy = energyBase()
+                currentProg = programExecutionQueue.popLast()!
+            }
+            return currentProg
+        } else {
+            return allIncludedPrograms.randomElement()!
+        }
+    }
+
+    private func regenProgramList() {
+        assert(programExecutionQueue.count == 0)
+        let edgeCounts = covEvaluator.getEdgeCounts()
+        var edgeCountsSorted = edgeCounts
+        edgeCountsSorted.sort()
+
+        // Find the edge with the smallest count
+        var startIndex = -1
+        for (i, val) in edgeCountsSorted.enumerated() {
+            if val != 0 {
+                startIndex = i
+                break
+            }
+        }
+        assert(startIndex != -1)
+        
+        // Find the nth interesting edge's count
+        let desiredEdgeCount = max(size / desiredSelectionProportion, 30)
+        let endIndex = min(startIndex + desiredEdgeCount, edgeCountsSorted.count - 1)
+        let maxEdgeCountToFind = edgeCountsSorted[endIndex]
+
+        // Find the n edges with counts <= maxEdgeCountToFind. Age them appropriately,
+        // to ensure non-deterministic samples are not continuously selected
+        for (i, val) in edgeCounts.enumerated() {
+            // Applies dropout on otherwise valid samples, to ensure variety between instances
+            // This will likely select some samples multiple times, which is acceptable as
+            // it is proportional to how many infrquently hit edges the sample has
+            if val != 0 && val <= maxEdgeCountToFind && probability(1 - dropoutRate) { 
+                if let prog = edgeMap[UInt32(i)] {
+                    programExecutionQueue.append(prog)
+                } else {
+                    logger.warning("Failed to find edge in map")
+                }
+            }
+        }
+        assert(programExecutionQueue.count > 0)
+        logger.info("Markov Corpus selected \(programExecutionQueue.count) new programs")
+    }
+
+    public func exportState() throws -> Data {
+        let res = try encodeProtobufCorpus(allIncludedPrograms)
+        logger.info("Successfully serialized \(allIncludedPrograms.count) programs")
+        return res
+    }
+    
+    public func importState(_ buffer: Data) throws {
+        let newPrograms = try decodeProtobufCorpus(buffer)        
```

Comment:
Doesn't this need to clear the coverage bitmap of the evaluator?

Reply:
Currently, the semantics of importState differ between BasicCorpus and MarkovCorpus. BasicCorpus removes its whole program list here: https://github.com/googleprojectzero/fuzzilli/blob/master/Sources/Fuzzilli/Core/Corpus.swift#L128, and resets all edges. The current implementation of MarkovCorpus just adds them individually.

I'm not very familiar with the networking code, but took a quick look. Am I right in thinking that importState is only called when importing a corpus from disk, or syncing with the master instance for the first time? If so, clearing the bitmap probably makes sense.

Reply:
Yes, correct. Currently the network code will run `importState` once, when connecting to the master and receiving the SYNC packet: https://github.com/googleprojectzero/fuzzilli/blob/master/Sources/Fuzzilli/Modules/NetworkSync.swift#L625

I think the correct behavior of `importState` should be to erase any local state (i.e. clear coverage, empty the corpus, etc.), then set the local state to the given state. So I think for the MarkovCorpus, that means clearing all coverage state and clearing all state of the MarkovCorpus, then import all programs in the corpus (which means executing them at least once to get the covered edges)

Reply:
Ok. That makes sense, and should be easy to implement. What do you think of adding a `clearState` function to CovEvaluator?

Reply:
So what's the status of this? I think adding a `clearState` or `reset` method to the CovEvaluator (or generally the evaluator) should be good.